File: 201-210/Count Primes.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cmath>

using namespace std;


class Solution {
public:
    int countPrimes(int n) {
        
        if(n < 3) return 0;
        
        int limit = sqrt(n);
        int r(n-2);
        
        bool *A = new bool[n+1];
        memset(A, true, n+1);
        
        
        for(int i = 2;i <= limit;i++)
        {
            if(A[i] == true)
            {
                for(int j = i*i;j < n;j += i)
                {
                    if(A[j] == true) r--;
                    A[j] = false;
                }
            }
            
        }
        return r;
    }
};

int main()
{
    Solution s;
    cout << s.countPrimes(2);
}

File: python/201-210/Count Primes.py
import math

class Solution:
    # @param {integer} n
    # @return {integer}
    def countPrimes(self, n):
        if n <= 2:
            return 0
        c = int(math.sqrt(n))
        b = [False] * n;
        
        for i in range(3, c+1):
            if not b[i]:
                j = i*i
                k = i << 1
                while j < n:
                    b[j] = True
                    j += k
        
        res = 1
        for i in range(3, n, 2):
            if not b[i]:
                res += 1
        return res